### PR TITLE
#393 Added notification pattern

### PR DESCRIPTION
**Pull request by DateNolan**
Added the notification design pattern along with tests and readme. However unfortunately had some issues with getting the uml diagram to auto generate, would appreciate some help with that.

I would also like to mention that this code is similar but seperate to NatalieSty's implementation of the same pattern, we both adapted our code from the same source and as such I would like to acknowledge their contribution to this pattern.

### CODE REVIEW

**Comment by vishhhhhhhh** [comment]
Looks good :)

File: notification/README.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+layout: pattern 
+title: com.iluwater.notification.Notification
+folder: notification 
+permalink: /patterns/notification/ 
+categories: Behavioural
+language: en 
+tags:
+- Decoupling
+- Presentation
+---
```
Comment (iluwatar) [inline notification/README.md:11]: Please check the updated yaml frontmatter requirements in https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute.

To the tags we could also add `Domain`

File: notification/README.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+```
+
+The form then processes the submission and returns these error messages to the user, showing our notification worked.
+
+output:
+```java
+17:19:17.902 [main] INFO com.iluwater.notification.RegisterWorkerForm - Error 1: Name is missing: ""
+17:19:17.909 [main] INFO com.iluwater.notification.RegisterWorkerForm - Error 2: Occupation is missing: ""
+17:19:17.909 [main] INFO com.iluwater.notification.RegisterWorkerForm - Error 4: Worker registered must be over 18: "2016-07-13"
+17:19:17.909 [main] INFO com.iluwater.notification.RegisterWorkerForm - Not registered, see errors
```
Comment (iluwatar) [inline notification/README.md:229]: Should be com.iluwatar, please check the package naming

File: notification/pom.xml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.16</version>
+        </dependency>
```
Comment (iluwatar) [inline notification/pom.xml:29]: Use the version from the parent pom.xml

File: notification/pom.xml [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                </configuration>
+                <executions>
+                    <execution>
+                        <phase>process-classes</phase>
+                        <goals>
+                            <goal>map</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
```
Comment (iluwatar) [inline notification/pom.xml:71]: This plugin is already present in the parent pom.xml, so no need to add it here

File: notification/src/main/java/com/iluwater/notification/App.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package com.iluwater.notification;
+
+import java.time.LocalDate;
+
+/**
+ * You can think of this application as filling in the form with information on a worker.
+ * This could include their name, occupation, and date of birth as an example.
+ * This is then submitted by the form for processing in the domain layer of our application.
+ */
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwater/notification/App.java:9]: In this section, please explain the pattern idea briefly and then describe how the example code implements it

File: notification/src/main/java/com/iluwater/notification/App.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  private App() {
+  }
+
+  /**
+   * Program entry point.
+   *
+   * @param args command line arguments, not used here
+   */
+  public static void main(String[] args) {
+    RegisterWorkerForm form = new RegisterWorkerForm(NAME, OCCUPATION, DATE_OF_BIRTH);
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwater/notification/App.java:27]: You can use `var` keyword for local variables

**Review by iluwatar — CHANGES_REQUESTED** [review]

**Comment by DateNolan** [comment]
@iluwatar Resolved all feedback given!

File: notification/pom.xml (hunk 3 above)
Comment (iluwatar) [reply notification/pom.xml]: the version could be left out completely to use the one from the parent, but it's not a big deal

Comment (iluwatar) [reply notification/pom.xml]: Oh, actually it is and causes a build error. 

```
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] 'dependencies.dependency.version' for org.projectlombok:lombok:jar must be a valid version but is '${lombok.version}'. @ line 28, column 22
```

Comment (iluwatar) [reply notification/pom.xml]: Please resolve this issue and we can proceed with the merge @DateNolan

**Comment by DateNolan** [comment]
@iluwatar  I believe I resolved the issue with versioning

File: notification/README.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
+--- 
+layout: pattern 
+title: Notification
+category: Behavioural
+language: en 
+tags:
+- Decoupling
+- Presentation
+- Domain
+---
```
Comment (iluwatar) [inline notification/README.md:10]: The frontmatter needs still adjustments. If you look at the requirements at https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute, you'll notice that there is no `layout` and `tags` should be renamed to `tag`.

File: notification/src/main/java/com/iluwater/RegisterWorkerDto.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+   *
+   * @param name Name of the worker
+   * @param occupation occupation of the worker
+   * @param dateOfBirth Date of Birth of the worker
+   */
+  public void setupWorkerDto(String name, String occupation, LocalDate dateOfBirth) {
+    this.name = name;
+    this.occupation = occupation;
+    this.dateOfBirth = dateOfBirth;
+  }
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwater/RegisterWorkerDto.java:58]: Perhaps these params could be set in the constructor?

File: notification/src/test/java/com/iluwater/DateOfBirthErrorTest.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+public class DateOfBirthErrorTest {
+  @Test
+  public void dateOfBirthTest() {
+    //Define our variables
+    String name = "valid name";
+    String occupation = "valid occupation";
+    LocalDate dateOfBirth = LocalDate.of(2016, 7, 13);
```
Comment (iluwatar) [inline notification/src/test/java/com/iluwater/DateOfBirthErrorTest.java:15]: Use `var` for local variables

Comment (iluwatar) [reply notification/src/test/java/com/iluwater/DateOfBirthErrorTest.java:15]: Check this elsewhere too

File: notification/src/main/java/com/iluwater/RegisterWorkerForm.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+@Slf4j
+public class RegisterWorkerForm {
+  String name;
+  String occupation;
+  LocalDate dateOfBirth;
+  RegisterWorkerDto worker;
+  /**
+   * Service super type which the form uses as part of its service layer.
+   */
+  RegisterWorkerService service = new RegisterWorkerService();
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwater/RegisterWorkerForm.java:20]: Is this constructor used anywhere?

File: notification/src/main/java/com/iluwater/RegisterWorkerForm.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+   *
+   * @param name name of worker
+   * @param occupation occupation of the worker
+   * @param dateOfBirth date of birth of the worker
+   */
+  public RegisterWorkerForm(String name, String occupation, LocalDate dateOfBirth) {
+    this.name = name;
+    this.occupation = occupation;
+    this.dateOfBirth = dateOfBirth;
+  }
```
Comment (iluwatar) [inline notification/src/main/java/com/iluwater/RegisterWorkerForm.java:33]: You can use Lombok to create this kind of constructors

**Review by iluwatar — CHANGES_REQUESTED** [review]

**Comment by iluwatar** [comment]
https://github.com/iluwatar/java-design-patterns/issues/393

**Pull request closed without merging**